### PR TITLE
Support parallel and add random search 

### PR DESCRIPTION
**Pull request by qingyun-wu**

### CODE REVIEW

File: flaml/automl.py [hunk 1]
```diff
@@ -1209,34 +1210,51 @@ def _search_parallel(self):
...
-            max_resource=self.max_resource,
-            config_constraints=[(partial(size, self._state), '<=', self._mem_thres)],
-            metric_constraints=self.metric_constraints)
-        search_alg = ConcurrencyLimiter(search_alg, self._n_concurrent_trials)
+        if self._hpo_method == 'random':
+            points_to_evaluate = self.points_to_evaluate.copy()
+            to_del = []
+            for k, v in self.search_space.items():
+                if not hasattr(v, 'is_valid'):
+                    to_del.append(k)
```
Comment (sonichi) [inline flaml/automl.py:1224]: What is 'is_valid‘? We don't have it in our search space

Comment (qingyun-wu) [reply flaml/automl.py:1224]: > checking whether the specific domain has 'is_valid'

Comment (sonichi) [reply flaml/automl.py:1224]: We never set 'is_valid' in our code, right?

File: test/test_automl.py [hunk 2]
```diff
@@ -478,6 +496,35 @@ def test_parallel_xgboost_random(self):
         # use random search as the hpo_method
         self.test_parallel_xgboost(hpo_method='random')
 
+    def test_random_out_of_memory(self):
+        automl_experiment = AutoML()
+        automl_experiment.add_learner(learner_name='large_lgbm', learner_class=MyLargeLGBM)
+        automl_settings = {
+            "time_budget": 10,
```
Comment (sonichi) [inline test/test_automl.py:503]: 2

File: flaml/automl.py [hunk 3]
```diff
@@ -943,10 +944,18 @@ def train(config: dict):
             if sample_size:
                 config['FLAML_sample_size'] = sample_size
             estimator = config['learner']
-            del config['learner']
-            result = states[estimator].training_function(config)
-            return result
-
+            # check memory constraints before training
+            if states[estimator].learner_class.size(config) <= MEM_THRES:
```
Comment (sonichi) [inline flaml/automl.py:948]: mem_thres is a configurable option with default value MEM_THRES

Comment (qingyun-wu) [reply flaml/automl.py]: fixed. Thanks!

File: test/test_automl.py (hunk 2 above)
Comment (qingyun-wu) [reply test/test_automl.py]: changed. thanks!

File: flaml/data.py [hunk 4]
```diff
@@ -141,14 +141,14 @@ def get_output_from_log(filename, time_budget):
...
-            time_used = record.total_search_time
+            time_used = record.wall_clock_time
             val_loss = record.validation_loss
             config = record.config
             learner = record.learner.split('_')[0]
             sample_size = record.sample_size
             train_loss = record.logged_metric
 
-            if time_used < time_budget:
+            if time_used < time_budget and val_loss != np.inf:
```
Comment (sonichi) [inline flaml/data.py:151]: np.isfinite(val_loss)

File: flaml/ml.py [hunk 5]
```diff
@@ -234,13 +236,13 @@ def evaluate_model_CV(
...
             eval_metric, task, labels, budget_per_train,
-            train_loss=train_loss, fit_kwargs=fit_kwargs)
+            log_training_metric=log_training_metric, fit_kwargs=fit_kwargs)
         if weight is not None:
             fit_kwargs['sample_weight'] = weight
         valid_fold_num += 1
         total_fold_num += 1
         total_val_loss += val_loss_i
-        if train_loss is not False or not isinstance(eval_metric, str):
+        if log_training_metric is not False or not isinstance(eval_metric, str):
```
Comment (sonichi) [inline flaml/ml.py:245]: remove 'is not False'

File: flaml/ml.py [hunk 6]
```diff
@@ -261,7 +263,7 @@ def evaluate_model_CV(
             break
     val_loss = np.max(val_loss_list)
     n = total_fold_num
-    if train_loss is not False or not isinstance(eval_metric, str):
+    if log_training_metric is not False or not isinstance(eval_metric, str):
```
Comment (sonichi) [inline flaml/ml.py:266]: remove 'is not False'

File: flaml/automl.py [hunk 7]
```diff
@@ -1488,6 +1623,25 @@ def _search(self):
                 logger.info(f'ensemble: {stacker}')
                 self._trained_estimator = stacker
                 self._trained_estimator.model = stacker
+            elif self._retrain_final:
+                # reset time budget for retraining
+                self._state.time_from_start -= self._state.time_budget
+                if (self._state.time_budget - self._state.time_from_start
```
Comment (qingyun-wu) [inline flaml/automl.py:1629]: why still need to check the budget?

File: .github/workflows/python-package.yml [hunk 8]
```diff
@@ -41,7 +41,7 @@ jobs:
       - name: If linux or mac, install ray
         if: (matrix.os == 'macOS-latest' || matrix.os == 'ubuntu-latest') && matrix.python-version != '3.9'
         run: |
-          pip install -e .[ray]
+          pip install -e .[ray,forecast]
```
[the remark this answers is not in the document]
Comment (sonichi) [reply .github/workflows/python-package.yml:44]: installation in windows fails

File: flaml/automl.py [hunk 9]
```diff
@@ -1032,62 +1032,71 @@ def custom_metric(
                         return metric_to_minimize, metrics_to_log
 
                 which returns a float number as the minimization objective,
-                and a tuple of floats or a dictionary as the metrics to log
+                and a tuple of floats or a dictionary as the metrics to log.
             task: A string of the task type, e.g.,
-                'classification', 'regression'
-            n_jobs: An integer of the number of threads for training
-            log_file_name: A string of the log file name
+                'classification', 'regression', 'forecast'.
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/automl.py:1037]: yes but I can't modify the other PR.

File: flaml/automl.py (hunk 7 above)
Comment (sonichi) [reply flaml/automl.py:1629]: to prevent too small budget used for retraining.

**Comment by sonichi** [comment]
close #171 #163

**Comment by sonichi** [comment]
#174

**Review by sonichi — APPROVED** [review]

**Pull request merged**